Introducing tech preview of the new AWX UI

TheRealHaoLiu opened the pull request:
##### SUMMARY
- Makefile file and make targets for pulling in and build new UI from https://github.com/ansible/ansible-ui
- Add banner switch between old UI and new UI
- Add new setting UI_NEXT to enable/disable new UI
- Add new config to display the setting for UI_NEXT (for non-admin)
- Add new path `/ui_next` to access the new UI

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - API
 - UI
 - Other

##### AWX VERSION
```
awx: 21.13.1.dev30+ge18a755f6b
```

##### TEST SCENARIOS
Test environment: fedora 37 VM setup using https://github.com/TheRealHaoLiu/fedora-setup/tree/main
docker-compose related  tests:
🟢 able to start API server without building ui or ui-next in docker-compose development environment (verify via access to /api/v2/ping)
🟢 able to build ui and ui-next on machine directly (verified via `make ui-devel && make ui-next`)
🟢 able to build ui and ui-next in docker-compose development environment (ui-next src from git)
🟢 after building ui without building ui-next with settings.UI_NEXT enabled verified that the placehold index page show up (telling people that UI_NEXT was not built yet)
🟢 after building ui and ui-next collectstatic is able to collect the static files successfully and ui show up

kube build related tests:
🟢 able to build awx with `make awx-kube-build` and old and new UI appears successfully (with spud logo)
🟢 CI feature branch built image function as expected with both old and new UI appear successfully (with spud logo)

release build related tests:
🟢 moc stage job was able to successfully build and push the moc release image https://github.com/ansible/awx/actions/runs/4443351332/jobs/7800534464
🟢 deployment using the image generated by stage CI job produced image is successful both old and new ui function (with AWX logo)

Function related tests:
🟢 /ui_next show 404 when settings.UI_NEXT = False
🟢 banner on old UI for switching to new UI does not appear when settings.UI_NEXT = False
🟢 banner on old UI for switching to new UI appear for normal user when settings.UI_NEXT = True

TheRealHaoLiu committed:
Build and serve UI_NEXT

- Add new makefile for building ui_next
- Add setting to toggle ui_next
- Add URL path for displaying ui_next
- Update collectstatic and template dir config to serve ui_next

kialam committed:
Add banner to dashboard page.

Co-Authored-By: kialam <2293210+kialam@users.noreply.github.com>

mabashian committed:
Adds UI bits for new UI_NEXT system setting

TheRealHaoLiu committed:
Add ui_next to /api/v2/config

- Add ui_next to /api/v2/config
- enable banner to show up for normal user since /api/v2/settings is only available to admin users

mabashian committed:
Only show tech preview banner when config.ui_next is true.  Use brandName variable in tech preview banner.

mabashian committed:
Fix use of brandName

TheRealHaoLiu committed:
throw 404 when UI_NEXT false

TheRealHaoLiu committed:
Refine UI_NEXT Makefile and update README

TheRealHaoLiu committed:
make dev-env test pass

mabashian committed:
Adds missing mock for fetching the brand name

kialam committed:
Fix brandname in banner.

TheRealHaoLiu committed:
add new UI icons

fosterseth commented on awx/main/conf.py:
```diff
@@ -795,6 +795,16 @@
+    field_class=fields.BooleanField,
+    default=False,
+    label=_('Enable Next Generation User Interface'),
+    help_text=_('Enable the next generation user interface.'),
```
optional suggestion:

"next generation" has other colloquial meanings, for example to indicate "cutting edge"

maybe we can consider "Enable experimental user interface"

fosterseth replied:
or "tech preview" as indicated elsewhere

TheRealHaoLiu replied:
updated

fosterseth commented on awx/ui/src/screens/Setting/shared/data.allSettingOptions.json:
```diff
@@ -639,6 +639,15 @@
+      "UI_NEXT": {
+        "type": "boolean",
+        "required": false,
+        "label": "Enable Next Generation User Interface",
```
should be "preview of new" here and a few other spots

TheRealHaoLiu replied:
updated

TheRealHaoLiu replied:
fixed

fosterseth reviewed (approved):
pulled it down and tested out the instructions in the readme. Was able to build new UI successfully, navigate to new UI, and tested out the banner functionality to flip between old and new UI

john-westcott-iv commented on Makefile:
```diff
@@ -449,7 +451,7 @@ HEADLESS ?= no
 dist/$(SDIST_TAR_FILE):
 else
-dist/$(SDIST_TAR_FILE): $(UI_BUILD_FLAG_FILE)
+dist/$(SDIST_TAR_FILE): $(UI_BUILD_FLAG_FILE) ui-next
```
Can we add a comment in here that ui-next is an imported target? It might save someone some time in the future.

jjwatt replied:
Yes, and a "MaybeImported" target at that :). Please leave a comment about that or else folks might be grepping for a `ui-next` recipe in this file and feeling hurt that it's not there.

TheRealHaoLiu replied:
added note to indicate that `ui-next` is imported

john-westcott-iv commented on awx/ui/public/static/media/192.png:
We are adding these files specifically for the new UI but then rebranding them as AWX on release, is that correct?

TheRealHaoLiu replied:
these file are for devel build for release build the icon come from https://github.com/ansible/awx-logos will override what's here

john-westcott-iv commented on awx/ui/src/screens/Setting/MiscSystem/MiscSystemDetail/MiscSystemDetail.test.js:
```diff
@@ -113,6 +114,7 @@ describe('<MiscSystemDetail />', () => {
     assertVariableDetail(wrapper, 'Remote Host Headers', '[]');
     assertVariableDetail(wrapper, 'Proxy IP Allowed List', '[]');
     assertDetail(wrapper, 'Global default execution environment', 'Foo');
+    assertDetail(wrapper, 'Enable Next Generation User Interface', 'On');
```
Did we want to change this value to be consistent with the wording in `awx/main/conf.py` (Enable Preview of New User Interface)?

TheRealHaoLiu replied:
fixed

john-westcott-iv commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,97 @@
+
+.PHONY: ui-next/build
+## Build ui-next/build
+ui-next/build: $(UI_NEXT_DIR)/build
```
Is this target necessary? Can ui-next just call `${UI_NEXT_DIR}/build` directly?

TheRealHaoLiu replied:
its not necessary...i added this for symetry...

john-westcott-iv commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,97 @@
+			git clone --depth 1 --branch $(UI_NEXT_GIT_BRANCH) $(UI_NEXT_GIT_REPO) $(UI_NEXT_DIR)/src || true; \
+		fi; \
+	else \
+		echo "SKIP: ui-next/src. UI_NEXT_LOCAL and UI_NEXT_GIT_REPO are not set."; \
```
Should we have an exit status code here since we can't actually make this target?

TheRealHaoLiu replied:
fixed now fail `make src` if UI_NEXT_LOCAL and UI_NEXT_GIT_REPO not set

john-westcott-iv commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,97 @@
+				echo "=== Updating $(UI_NEXT_DIR)/src from $(UI_NEXT_GIT_BRANCH) of $(UI_NEXT_GIT_REPO) ==="; \
+				git fetch && git pull; \
+			else \
+				echo "=== Cloning $(UI_NEXT_DIR)/src from $(UI_NEXT_GIT_BRANCH) of $(UI_NEXT_GIT_REPO) ==="; \
```
This and the outer else feels like it could be a little reworked to reduce redundency. Can the if above just `exit 0` and then thise else just `rm /src` and then remove the lower else so that it just echo's and clones? (or something along those lines).

jjwatt replied:
Again, echoing John's comment: this has warts. The first code smell is that it's just too long. In fact, that would cause `make` to fail in some environments because of argument list length restrictions! Make essentially passes everything as one long line as arguments to `sh -c`. Yes, it could use reworking, but even then it's probably doing too much. My first thought is just *lift*. Lift it to a helper script, possibly with params--you can always pass make macro vars `$()` into the helper script as-needed. Then, I don't even care that it's not great script, and it will be more likely to be improved later, will be picked up by shell linters automatically, etc. As an alternative to that, we could probably write much of the functionality in actual Make code.
That's a lie, tho. My first thought is really to let it in if it works well enough :)
But, if and when we want to fix it, we need to simplify it here and/or move it to a script.

TheRealHaoLiu replied:
as discussed previously will address this outside of this PR...yes i hate this blob of bash too

TheRealHaoLiu replied:
added TODO to address this later

john-westcott-iv commented on awx/ui_next/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Set src of the ui_next repo
+
+### via GIT
```
Since the Makefile will use the default repo could we say something like `from a custom GIT repo`

john-westcott-iv replied:
Also, it looks like a bad find and replace happened. This should be `UI_NEXT_GIT_REPO` and the branch should just be `UI_NEXT_GIT_BRANCH`

TheRealHaoLiu replied:
fixed

john-westcott-iv commented on awx/ui_next/README.md:
````diff
@@ -0,0 +1,39 @@
+export UI_NEXT_GIT_BRANCH_BRANCH=main
+```
+
+### via symlink to existing clone
````
Can we say something here about which has prescidence if both UI_NEXT_GIT_BRANCH_REPO and UI_NEXT_LOCAL are both set?

TheRealHaoLiu replied:
fixed

john-westcott-iv reviewed (commented):
Should we add/change anything in https://github.com/ansible/awx/blob/devel/awx/ui/CONTRIBUTING.md?

john-westcott-iv commented:
When you turn on/off the setting in the UI settings page and then go back to the dashboard view the message does not toggle until you do a full reload of the page. Is there a way to have it re-check if the header should be on/off on a dashboard load?

john-westcott-iv commented:
I put on my user hat and went through the default motions:
```
clone repo
make ui-devel
make docker-compose-build
make docker-compose
```

When I click on the link to load the new dashboard I get a 500 because the template does not exist for the view:
```
TemplateDoesNotExist at /ui_next/dashboard/
index_awx.html
```

Can we add a default template that has a message like "Did you forget to build the UI Next interface"?

john-westcott-iv committed:
Adding default index_awx.html incase user forgets to build ui-next

john-westcott-iv commented:
In the blue banner in UI Next it references `Ansible Automation Platform` should that be changed to AWX for community stuff?

john-westcott-iv commented:
I had awx running and then did `make ui-next` and ended up with a blank screen until I redid `make docker-compose-build && make docker-compose`. Is that expected? I'm thinking yes because the index file was there to serve the HTML but collectstatic needs to be re-run to pickup all of the additional files. If thats accurate can we add it to whatever build documentation we are going to (hopefully) add?

john-westcott-iv commented on tools/docker-compose/bootstrap_development.sh:
```diff
@@ -25,6 +25,7 @@ fi
 
 # Make sure that the UI static file directory exists, Django complains otherwise.
 mkdir -p /awx_devel/awx/ui/build/static
+mkdir -p /awx_devel/awx/ui_next/build/awx
```
Since we added a default template, I think this line should no longer be needed.

TheRealHaoLiu replied:
i ended up moving the placeholder index_awx.html out because 
![image](https://user-images.githubusercontent.com/44379968/226410678-333dfada-a239-419a-8e52-8c6e23cbea29.png)

but great idea and thanks for the commit

kialam committed:
Commit .po files.

jjwatt commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,97 @@
+	@echo "=== Setting up $(UI_NEXT_DIR)/src ==="
+	@if [ ! -z "$(UI_NEXT_LOCAL)" ]; then \
+		if [ -d $(UI_NEXT_DIR)/src ]; then \
+			if [ "$$(readlink $(UI_NEXT_DIR)/src)" = "$(UI_NEXT_LOCAL)" ]; then \
```
Let's try not to use `readlink` so we're inclusive of BSD (and macOS, I guess).

TheRealHaoLiu replied:
this one works for macOS... i verified with BSD readlink

jjwatt commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,97 @@
+ui-next/build: $(UI_NEXT_DIR)/build
+
+## Real target for ui-next.
+$(UI_NEXT_DIR)/build: ui-next/src/build
```
Somewhat in line with John's comment: It's not always clear which of these targets are meant to be PHONY and which are meant to be real targets. And, more importantly to functionality:
1. Real targets should probably never depend on phony targets because PHONY targets will always evaluate as needing updating, and so they'll make the "real" targets always need updating, too, since the real targets *depend* on the always-out-of-date PHONY target, and if something might have changed with it, then something might have changed with our target and...
2. Real directories should almost never be real targets
Similar, but distinct reasoning: A POSIX directory is "updated" or "changed" anytime any metadata about it changes (!), so that would include any files atime (on certain filesystems), new/removed directory, etc. etc. and GNU Make looks at that as a trigger to update the target, so directories will often need updating.
It's not immediately super clear to me what the intent is, so I'm not sure what to recommend instead, but it's late on a Friday. I wouldn't block merging this if it works, but would definitely want to rewrite a lot of it in a followup commit.

TheRealHaoLiu replied:
explicitly PHONY the make targets

TheRealHaoLiu committed:
Move placeholder index_awx.html out of build dir

- move placeholder index_awx.html out of ui_next build dir
- copy index_awx.html to build dir during development bootstrap if UI_NEXT has not been build

TheRealHaoLiu committed:
PHONY all UI_NEXT build target

- they were all PHONY to start with and also all target are written to be rerun able

TheRealHaoLiu committed:
fail UI_NEXT make src if variable not set

TheRealHaoLiu committed:
Update UI_NEXT README

also cleanup some small things

TheRealHaoLiu committed:
Update UI_NEXT help text in UI

TheRealHaoLiu committed:
Add note to indicate ui-next is imported target

kialam commented:
> In the blue banner in UI Next it references `Ansible Automation Platform` should that be changed to AWX for community stuff?

Yes, those changes have already been made.

kialam commented:
> When you turn on/off the setting in the UI settings page and then go back to the dashboard view the message does not toggle until you do a full reload of the page. Is there a way to have it re-check if the header should be on/off on a dashboard load?

It would require another API call to get /config in order to refresh the banner. The banner doesn't auto update unless a refresh is called for now, but can revisit this.

TheRealHaoLiu commented on .gitignore:
```diff
@@ -157,7 +157,11 @@ use_dev_supervisor.txt
 *.unison.tmp
 *.#
 /awx/ui/.ui-built
+/awx/ui_next/.ui-built
```
remove this line that file is no longer needed

TheRealHaoLiu commented on awx/ui_next/Makefile:
```diff
@@ -0,0 +1,102 @@
+## UI_NEXT_DIR: Relative path to the directory containing this Makefile
+UI_NEXT_DIR := $(patsubst %/,%,$(dir $(lastword $(MAKEFILE_LIST))))
+
+# ## Path to your local clone of the UI_NEXT repo
```
remove #12627
